### PR TITLE
Multithread ETH sync

### PR DESCRIPTION
**Pull request by kajoseph**
**NOTE: This PR will bump the minimum required node version from 8 to 10 for the use of worker threads.**

This PR:

- Adds multi-threaded and multi-client syncing
- Introduces a GethRPC class. This is not a properly working class and thus has a throw in the constructor to prevent usage at this time. The full support for geth will be added in a subsequent PR
- Refactors Parity naming to Erigon
- Refactors some typings for consistency
- Updates a few miscellaneous things:
  - `bitcore-node.service` file to use typescript outdir
  - removes unused and outdated `CHANGELOG.md`
  - fixes Web3Proxy api endpoint and adds config value
  - updates karma deps in bitcore-build
  
TODOs:
- geth support
- DRY-ing up some code, specifically the processBlock functions in p2p.ts and syncWorker.ts

### CODE REVIEW

File: packages/bitcore-node/src/modules/ethereum/types.ts [hunk 1]
```diff
@@ -175,8 +177,8 @@ export interface EthTransactionJSON {
   nonce: number;
   to: string;
   from: string;
-  abiType?: IEthTransaction['abiType'];
-  decodedData?: AbiDecodedData;
+  abiType?: IAbiDecodedData;
+  decodedData?: IAbiDecodedData;
```
Comment (escottalexander) [inline packages/bitcore-node/src/modules/ethereum/types.ts:181]: Seems like we should rework this to only use either abiType or decodedData. That should probably be a separate PR though.

File: packages/bitcore-node/src/modules/ethereum/models/block.ts [hunk 2]
```diff
@@ -178,6 +181,50 @@ export class EthBlockModel extends BaseBlock<IEthBlock> {
...
+          return resolve([]);
+        }
+
+        const stream = self.collection
+          .find({ chain, network, height: { $gte: startHeight } })
+          .sort({ chain: 1, network: 1, height: 1 })
+          .addCursorFlag('noCursorTimeout', true);
+
+        const maxHeight = maxBlock.height;
+        let block = (await stream.next()) as IEthBlock;
```
Comment (escottalexander) [inline packages/bitcore-node/src/modules/ethereum/models/block.ts:201]: 🤯

File: packages/bitcore-node/src/modules/ethereum/models/block.ts [hunk 3]
```diff
@@ -178,6 +181,50 @@ export class EthBlockModel extends BaseBlock<IEthBlock> {
     }
     return JSON.stringify(transform);
   }
+
+  async verifySyncHeight(params: { chain: string; network: string; startHeight?: number }): Promise<number[]> {
```
Comment (escottalexander) [inline packages/bitcore-node/src/modules/ethereum/models/block.ts:185]: Seems like this could be more accurately named "getUnsyncedBlocks" or something like that. I see later you call it "gaps". Maybe "getSyncedBlockGaps".

File: packages/bitcore-node/src/modules/ethereum/p2p/rpcs/index.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import Web3 from 'web3';
+import { ErigonRPC } from './erigonRpc';
+import { GethRPC } from './gethRpc';
+
+export const Rpcs = {
```
Comment (escottalexander) [inline packages/bitcore-node/src/modules/ethereum/p2p/rpcs/index.ts:5]: Really digging this design choice.

**Review by escottalexander — DISMISSED** [review]
This looks good to me. I tested on my local and it works good as well. Nice job with all the refactoring and type-adding. Left a couple of suggestions but they are somewhat personal preference.

File: packages/bitcore-node/src/modules/ethereum/types.ts (hunk 1 above)
Comment (kajoseph) [reply packages/bitcore-node/src/modules/ethereum/types.ts:181]: Agreed.

**Comment by kajoseph** [comment]
I merged in https://github.com/bitpay/bitcore/pull/3349 to ensure no merge conflicts. Since it and this will require Node version increases, and thus a major version bump, they should be merged into master at the same time.

File: packages/bitcore-node/src/modules/ethereum/p2p/p2p.ts [hunk 5]
```diff
@@ -116,13 +126,13 @@ export class EthP2pWorker extends BaseP2PWorker<IEthBlock> {
...
     let disconnected = false;
-    const { host, port } = this.chainConfig.provider;
+    const { host, port } = this.chainConfig.provider || this.chainConfig.providers![0];
     while (!this.disconnecting && !this.stopping) {
       try {
         if (!this.web3) {
           const { web3 } = await this.getWeb3();
           this.web3 = web3;
-          this.rpc = new ParityRPC(this.web3);
+          this.rpc = new Rpcs[this.chainConfig.client || 'erigon'](this.web3);
```
Comment (escottalexander) [inline packages/bitcore-node/src/modules/ethereum/p2p/p2p.ts:135]: Because we say we don't maintain Erigon support, once we have the Geth class in working order we should switch this to default to Geth.

**Review by escottalexander — DISMISSED** [review]
Looks good!

File: packages/bitcore-node/src/modules/ethereum/models/block.ts [hunk 6]
```diff
@@ -17,6 +17,7 @@ export class EthBlockModel extends BaseBlock<IEthBlock> {
 
   async onConnect() {
     super.onConnect();
+    this.collection.createIndex({ chain: 1, network: 1, height: 1 }, { background: true });
```
Comment (nitsujlangston) [inline packages/bitcore-node/src/modules/ethereum/models/block.ts:20]: we already have 
```
chain: 1, network: 1, processed: 1, height: -1
```

File: packages/bitcore-node/src/modules/ethereum/p2p/p2p.ts [hunk 7]
```diff
@@ -1,31 +1,35 @@
...
 import { BaseP2PWorker } from '../../../services/p2p';
+import { IEthNetworkConfig } from '../../../types/Config';
 import { valueOrDefault } from '../../../utils/check';
 import { wait } from '../../../utils/wait';
 import { ETHStateProvider } from '../api/csp';
 import { EthBlockModel, EthBlockStorage } from '../models/block';
 import { EthTransactionModel, EthTransactionStorage } from '../models/transaction';
-import { IEthBlock, IEthTransaction, ParityBlock, ParityTransaction } from '../types';
-import { ParityRPC } from './parityRpc';
+import { ErigonBlock, ErigonTransaction, IEthBlock, IEthTransaction } from '../types';
```
Comment (nitsujlangston) [inline packages/bitcore-node/src/modules/ethereum/p2p/p2p.ts:15]: did you mean to call this erigon?

File: packages/bitcore-node/src/modules/ethereum/p2p/sync.ts [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        // TODO account for this.syncQueue
+        const oneSecond = 1000;
+        const blocksProcessed = this.currentHeight - startHeight;
+        const elapsedMinutes = (Date.now() - startTime) / (60 * oneSecond);
+        logger.info(
+          `${timestamp()} | Syncing... | Chain: ${chain} | Network: ${network} |${(blocksProcessed / elapsedMinutes)
+            .toFixed(2)
+            .padStart(8)} blocks/min | Height: ${this.currentHeight.toString().padStart(7)}`
+        );
+      }, 1000);
```
Comment (nitsujlangston) [inline packages/bitcore-node/src/modules/ethereum/p2p/sync.ts:85]: can call this oneSecond, as above ;)

File: packages/bitcore-node/src/modules/ethereum/p2p/p2p.ts (hunk 7 above)
Comment (kajoseph) [reply packages/bitcore-node/src/modules/ethereum/p2p/p2p.ts:15]: Yes, I segregated the Parity/Erigon unifying to the `gethSupport` (#3407)  PR. I renamed them in there

File: packages/bitcore-node/src/modules/ethereum/p2p/sync.ts (hunk 8 above)
Comment (kajoseph) [reply packages/bitcore-node/src/modules/ethereum/p2p/sync.ts]: ;)

File: packages/bitcore-node/src/modules/ethereum/models/block.ts (hunk 6 above)
Comment (kajoseph) [reply packages/bitcore-node/src/modules/ethereum/models/block.ts]: adjusted queries to to force use existing index

**Review by nitsujlangston — DISMISSED** [review]

**Review by nitsujlangston — APPROVED** [review]

**Review by nitsujlangston — APPROVED** [review]
